pilot/pkg/serviceregistry/kube/mcs/discoverability.go

```diff
@@ -0,0 +1,310 @@
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/queue"
+	"istio.io/pkg/log"
+	"istio.io/pkg/monitoring"
+)
+
+const (
+	discoverabilityLogPrefix = "MCS DiscoverabilityController:"
+)
+
+var (
+	empty = struct{}{}
+
+	eventTag   = monitoring.MustCreateLabel("event")
+	clusterTag = monitoring.MustCreateLabel("cluster")
+
+	serviceExportEvents = monitoring.NewSum(
+		"service_export_events",
+		"MCS ServiceExport events fromm all k8s clusters.",
+		monitoring.WithLabels(eventTag, clusterTag),
+	)
+)
+
+var _ model.ServiceDiscoverabilityPolicy = &DiscoverabilityController{}
+
+func init() {
+	monitoring.MustRegister(serviceExportEvents)
+}
+
+func incrementEvent(event string, cluster kube.ClusterName) {
+	serviceExportEvents.With(eventTag.Value(event), clusterTag.Value(string(cluster))).Increment()
+}
+
+type DiscoverabilityConfig struct {
+	Clusters     *kube.ClusterSet
+	XDSUpdater   model.XDSUpdater
+	DomainSuffix string
+}
+
+// NewDiscoverabilityController creates a new DiscoverabilityController that observes the given ClusterSet.
+func NewDiscoverabilityController(cfg DiscoverabilityConfig) *DiscoverabilityController {
+	c := &DiscoverabilityController{
+		clusters:     cfg.Clusters,
+		xdsUpdater:   cfg.XDSUpdater,
+		domainSuffix: cfg.DomainSuffix,
+		exports:      make(map[kube.ClusterName]*exportedServices),
+		queue:        queue.NewQueue(time.Second),
+	}
+
+	c.clusters.AddClusterLifecycleListener(func(e kube.ClusterEvent, name kube.ClusterName) {
+		switch e {
+		case kube.ClusterAdded:
+			c.onClusterAdded(name)
+		case kube.ClusterDeleted:
+			c.onClusterDeleted(name)
+		}
+	})
+	return c
+}
+
+// DiscoverabilityController is a ServiceDiscoverabilityPolicy that is driven by MCS resources in each cluster.
+type DiscoverabilityController struct {
+	clusters     *kube.ClusterSet
+	xdsUpdater   model.XDSUpdater
+	domainSuffix string
+	queue        queue.Instance
+	exports      map[kube.ClusterName]*exportedServices
+
+	// mutex for access to the exports map. Updates to cluster and ServiceExports
+	// should be infrequent enough that we can use a single lock for the entire map.
+	mutex sync.RWMutex
+}
+
+// Run this controller.
+func (c *DiscoverabilityController) Run(stopCh <-chan struct{}) error {
+	log.Debugf("%s started", discoverabilityLogPrefix)
+	go c.queue.Run(stopCh)
+	return nil
+}
+
+func (c *DiscoverabilityController) onClusterAdded(name kube.ClusterName) {
+	c.queue.Push(func() error {
+		log.Debugf("%s adding cluster %s", discoverabilityLogPrefix, name)
+
+		cluster := c.clusters.GetClusterWithName(name)
+		if cluster == nil {
+			log.Warnf("%s failed adding cluster %s: cluster not found", discoverabilityLogPrefix, name)
+			return nil
+		}
+
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+
+		// Create the informer and register callbacks.
+		informer := cluster.Client().MCSApisInformer().Multicluster().V1alpha1().ServiceExports().Informer()
```

Comment:
where do we call .Run on the informer?

Reply:
It's using a sharedIndexInformer which is run when `client.Run` is called.